### ISSUE TITLE
Boolean MySQL migration vs. auto-model-attribute-typecast inconsistency

### ISSUE THREAD

**Issue by karelvasicek** [yiisoft/yii2#19808]
- `TINYINT(1)` column is created using `$this->boolean()` in migration
- `BIT(1)` column is type-casted to `bool` during model population (`TINYINT(1)` column is type-casted as `integer`)

**Comment by terabytesoftw** [comment]
@rob006 @bizley What would be the best way to solve this problem.

**Comment by bizley** [comment]
I would check typecasting process.

**Comment by rob006** [comment]
Maybe we should have the same condition in MySQL `Schema` as for SQLite:

https://github.com/yiisoft/yii2/blob/5f3d36ea21dfc7b39d1c438e85971f03bf43f193/framework/db/sqlite/Schema.php#L333

**Comment by terabytesoftw** [comment]
> Maybe we should have the same condition in MySQL `Schema` as for SQLite:
> 
> https://github.com/yiisoft/yii2/blob/5f3d36ea21dfc7b39d1c438e85971f03bf43f193/framework/db/sqlite/Schema.php#L333

The change seems correct to me, but it will be BC, since it currently returns integer, i think it would be correct, and the other dbms return integer too, It would also be correct to normalize it.

https://github.com/yiisoft/yii2/blob/5f3d36ea21dfc7b39d1c438e85971f03bf43f193/tests/framework/db/SchemaTest.php#L431,

**Comment by VladimirRakovich** [comment]
If you look at the mysql documentation, you can see that this data type can be -128...127 signed and 0...255 unsigned. tinyint(1) doesn't equal the boolean type and this change will potentially break a lot of projects

[https://dev.mysql.com/doc/refman/8.0/en/integer-types.html](https://github.com/yiisoft/yii2-gii/issues/url)

**Comment by rob006** [comment]
Why would you use `tinyint(1)` if you want just `tinyint`? Defining length does not create any additional constrains and does not make data more compact.

**Comment by terabytesoftw** [comment]
Document https://github.com/yiisoft/yii2/blob/6804fbeae8aa5f8ad5066b50f1864eb0b9d77849/framework/db/QueryBuilder.php#L1183

**Comment by schmunk42** [comment]
I also stumbled upon this ... and I asked myself the following question: 

Why did `$this->boolean()` in migrations not simply create a `BIT(1)` column? Which was always typecasted to `boolean`.

**Comment by schmunk42** [comment]
> Why would you use `tinyint(1)` if you want just `tinyint`? Defining length does not create any additional constrains and does not make data more compact.

For new projects this would be fine, but I think @VladimirRakovich comment was about existing projects, where model values are now `boolean` instead of `integer` - this should be a concern, especially with PHP 8.1 and above.

**Comment by schmunk42** [comment]
@samdark 

> As for doing it at schema level, it may break things.

https://github.com/yiisoft/yii2-gii/issues/440#issuecomment-788069836

Why did you change your mind :) ?

**Comment by bizley** [comment]
I think this is because of MySQL docs where it was mentioned as a preferred implementation of boolean column (I hope I remember it correctly and not making it up).

**Comment by terabytesoftw** [comment]
This is an old bug, which was fixed according to the Yii2 documentation, so it is complicated to make any corrections to the master branch.

**Comment by rob006** [comment]
> For new projects this would be fine, but I think @VladimirRakovich comment was about existing projects

My questions stands also for existing projects - why would you specify length as `1` for tinyint? It does not have any practical effect, and for me it was always used to indicate that this is a boolean column.

**Comment by schmunk42** [comment]
> My questions stands also for existing projects - why would you specify length as 1 for tinyint? It does not have any practical effect, and for me it was always used to indicate that this is a boolean column.

Yeah, but this also works the other way :)
Since it has no practical effect why bother; or maybe you work against an existing database where this is actually used to store values from `0, 255`. Such an application would be completely broken.

**Comment by schmunk42** [comment]
> I think this is because of MySQL docs where it was mentioned as a preferred implementation of boolean column (I hope I remember it correctly and not making it up).

It's here: https://dev.mysql.com/doc/refman/8.0/en/numeric-type-syntax.html

But the alias does not really apply vice-versa.

**Comment by rob006** [comment]
Do we have a way to overwrite this detected type in somewhat convenient manner? I still think that `tinyint(1)` should be treated as bool by default, but it would be worth to add a way to overwrite detected type and handle edge cases when `tinyint(1)` actually represents a number.

**Comment by handcode** [comment]
IMHO it is simply wrong to assume in the schema that EVERY tinyint col (no matter if with or without length) is a boolean.
https://github.com/terabytesoftw/yii2/blob/b40de6a5380819472f5bde6c5d120e9c83146c87/framework/db/mysql/Schema.php#L282

this is valid (My)SQL and the `state` column is clearly NOT a boolean.

```
CREATE TABLE `dev1` (
       `id` int(11) NOT NULL,
       `state` tinyint(1) unsigned DEFAULT 0,
       PRIMARY KEY (`id`)
  ) ENGINE=InnoDB;

INSERT INTO `dev1` (`id`, `state`) VALUES (1, 255);

select * from `dev1`;
+----+-------+
| id | state |
+----+-------+
|  1 |   255 |
+----+-------+

```
And for MySQL the length param for tinyint has no effect. The possible range of values is only dependent on signed or unsigned.

**Comment by rob006** [comment]
> If you are right in this case, we can fix it

How? Column schema does not check what values are stored in this column.

**Comment by terabytesoftw** [comment]
> > If you are right in this case, we can fix it
> 
> How? Column schema does not check what values are stored in this column.

https://github.com/yiisoft/yii2/blob/6804fbeae8aa5f8ad5066b50f1864eb0b9d77849/framework/db/mysql/Schema.php#L282

**Comment by handcode** [comment]
> > If you are right in this case, we can fix it
> 
> How? Column schema does not check what values are stored in this column.

And even if you look at what is stored so far, how is schema supposed to know what i want to store tomorrow?

**Comment by rob006** [comment]
That is why I propose a way to define types of columns in config and keep current type detection as a fallback. For me this guessing  PHP type from DB schema was always a bit too magic, I would prefer to keep these definitions as part of the code with explicitly defined defaults or types.

**Comment by schmunk42** [comment]
Example and discussion from a non-PHP project: https://github.com/metabase/metabase/issues/3506

tldr: They implemented a driver flag `tinyInt1IsBit=false` where `TINYINT(1)` columns treated as numbers instead of booleans.

We have the manual saying `TINYINT(1)` and `BOOLEAN` are synonyms, although it is not enforced on db-level.
So you could argue, that using `TINYINT(1)` for values other than `0` and `1` is "wrong" - but yeah, this is more a made-up argument.

**Comment by terabytesoftw** [comment]
I just did a test for the boolean column, it is a Yii2 mapping, which phpTypeCast will always return a boolean value, be it tinyint(1) or bit(1), in any case as I said before,

**Comment by schmunk42** [comment]
Just for the record ... I did some tests on the db-level.

![Bildschirmfoto vom 2023-10-31 15-16-52](https://github.com/yiisoft/yii2/assets/649031/2a05e433-83e9-4f5e-925c-0ecb9ccdb4ae)


```
column       length
-------------------------
tiny1        1
tinynolen    4
tiny5        5
tiny0        4
``` 

`tinyint` and `tinyint(0)` is `tinyint(4)`

**Comment by handcode** [comment]
@terabytesoftw 
sorry, but if it is `boolean` or not has nothing to do with signed or unsigned. the signed/unsigned flag "just" define the possible range for tinyint.
https://dev.mysql.com/doc/refman/8.0/en/integer-types.html

Still valid SQL but still NO boolean col:
```
CREATE TABLE `dev1` (
       `id` int(11) NOT NULL,
       `state` tinyint(1) DEFAULT 0,
       PRIMARY KEY (`id`)
  ) ENGINE=InnoDB;

INSERT INTO `dev1` (`id`, `state`) VALUES (1, 127);

select * from dev1;
+----+-------+
| id | state |
+----+-------+
|  1 |   127 |
+----+-------+
```

To be clear: i understand the problem, that MySQL has no real boolean column type, and that the docs state that boolean "is" a tinyint(1), but the assumption the other way around, that every tinyint(1) **is** a boolean is simply wrong.

**Comment by terabytesoftw** [comment]
related:

https://github.com/yiisoft/yii2/pull/9339
https://github.com/yiisoft/yii2/pull/9324
https://github.com/yiisoft/yii2/issues/15471
https://github.com/yiisoft/yii2/pull/9324#issuecomment-128127967

Yii2 not only supports Mysql 8, it also supports the legacy versions.

**Comment by schmunk42** [comment]
> That is why I propose a way to define types of columns in config and keep current type detection as a fallback. For me this guessing PHP type from DB schema was always a bit too magic, I would prefer to keep these definitions as part of the code with explicitly defined defaults or types.

@rob006 How would that look like?

Also related in general:
- https://github.com/propelorm/Propel2/issues/1490

**Comment by VladimirRakovich** [comment]
> related:
> 
> #9339 #9324 #15471 [#9324 (comment)](https://github.com/yiisoft/yii2/pull/9324#issuecomment-128127967)

https://github.com/yiisoft/yii2/pull/9339#issuecomment-485972726
:slightly_smiling_face:

**Comment by terabytesoftw** [comment]
btw in Yii3 we use bit(1), but that here would be BC.

**Comment by handcode** [comment]
>  if it is correct that tinyint(1) would be synonymous with boolean, but not tinyint(1) unsigned which would be integer, then for me two solutions.
> 

- boolean is "stored" as integers 0|1 in a tinyint col because there is no real boolean type
- tinyint signed or unsigned are integers **not** booleans

see: https://github.com/yiisoft/yii2/issues/19808#issuecomment-1787325152

Why here, because of this one word _synonyms_ from the doc, it is pretended that there is a defined column data type `Boolean`, and that this `Boolean` type is `tinyint(1)`, I honestly don't understand. MySQL has no defined storage type for boolean. If there would be one, it would surely be called `boolean` and stored as `boolean` and it would not be allowed to store an integer 42 in this col.

Just open a MySQL and try it yourself what you can write into a tiniyint field and what you get out of it. 

If you still think it's a good idea to **always** cast the value from a tinyint(1) column to the php type boolean, go ahead. 

This would be obviously a bug and will potentially break perfectly valid code, but hey, we would have a _boolean_ or at least we could pretend that it is so.

I'm out.

**Comment by terabytesoftw** [comment]
> > if it is correct that tinyint(1) would be synonymous with boolean, but not tinyint(1) unsigned which would be integer, then for me two solutions.
> 
> * boolean is "stored" as integers 0|1 in a tinyint col because there is no real boolean type
> * tinyint signed or unsigned are integers **not** booleans
> 
> see: [#19808 (comment)](https://github.com/yiisoft/yii2/issues/19808#issuecomment-1787325152)
> 
> Why here, because of this one word _synonyms_ from the doc, it is pretended that there is a defined column data type `Boolean`, and that this `Boolean` type is `tinyint(1)`, I honestly don't understand. MySQL has no defined storage type for boolean. If there would be one, it would surely be called `boolean` and stored as `boolean` and it would not be allowed to store an integer 42 in this col.
> 
> Just open a MySQL and try it yourself what you can write into a tiniyint field and what you get out of it.
> 
> If you still think it's a good idea to **always** cast the value from a tinyint(1) column to the php type boolean, go ahead.
> 
> This would be obviously a bug and will potentially break perfectly valid code, but hey, we would have a _boolean_ or at least we could pretend that it is so.
> 
> I'm out.

The first thing I must clarify is that I have not said that there is a boolean type in MYSQL, simply by not having a specific type a mapping is made, and this is the Framework specification, you can consult QueryBuilder in getColumnType, if it is correct or not, It's another discussion, it simply refers me to the existing documentation, keep in mind that this also handles legacy versions of mysql, where in the previous documentation it said that tinyint(1) was synonymous with boolean, i'm not making it up.

![image](https://github.com/yiisoft/yii2/assets/42547589/e7376199-065d-4aef-b043-39b677e63025)

**Comment by schmunk42** [comment]
@terabytesoftw 

> If you do the test, using the Framework, you will see that it only saves 0 and 1, when you use $this->boolean().

How would the framework do that? Aren't you able to set `$model->is_something = 5` and save that.

In a project I have columns defined in migration with `$this->boolean()` and my expectation was that I can easily work with `true` and `false` - but that is not the case, firstly `gii` generates an integer validator, which prevents me from setting `true/false`, secondly values from the db are also integersin the model.

**Comment by terabytesoftw** [comment]
> @terabytesoftw
> 
> > If you do the test, using the Framework, you will see that it only saves 0 and 1, when you use $this->boolean().
> 
> How would the framework do that? Aren't you able to set `$model->is_something = 5` and save that.
> 
> In a project I have columns defined in migration with `$this->boolean()` and my expectation was that I can easily work with `true` and `false` - but that is not the case, firstly `gii` generates an integer validator, which prevents me from setting `true/false`, secondly values from the db are also integersin the model.

The framework works this way in this case, you can read the code and documentation, where it clearly says that tinyint(1) works with 0 for false, 1 for true, the framework does the typeCast, this is how it works for this case, In Yii3 to eliminate ambiguity we use bit(1), and that is why I asked here how the best way to solve it would be, since here applying bit(1) would be BC, but it eliminates the ambiguity.

**Comment by terabytesoftw** [comment]
https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-19.html

```
The type is . MySQL Connectors make the assumption that columns originated as columns; this exception enables them to continue to make that assumption. TINYINT(1)TINYINT(1)BOOLEAN
```

**Comment by schmunk42** [comment]
@samdark @bizley IMHO, the changes in #20045 should be reverted, there's just too much what could break.

Ie. all models generated with `gii` which have a property for a `tinyint(1)` column have `IntegerValidator`s, but the value would now be a `boolean`. And this is just one example.

**Comment by terabytesoftw** [comment]
Yes, i agree, this commit should be reverted.
